service/history/task/task_rate_limiter.go

```diff
@@ -0,0 +1,119 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+//go:generate mockgen -package $GOPACKAGE -destination task_rate_limiter_mock.go github.com/uber/cadence/service/history/task RateLimiter
+
+package task
+
+import (
+	"context"
+
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/quotas"
+	"github.com/uber/cadence/service/history/config"
+	"github.com/uber/cadence/service/history/shard"
+)
+
+type (
+	RateLimiter interface {
+		Allow(Task) bool
+		Wait(context.Context, Task) error
+	}
+
+	taskRateLimiterImpl struct {
+		logger           log.Logger
+		metricsScope     metrics.Scope
+		domainCache      cache.DomainCache
+		limiters         quotas.ICollection
+		enabled          dynamicconfig.BoolPropertyFn
+		enableShadowMode dynamicconfig.BoolPropertyFnWithDomainFilter
+	}
+)
+
+func NewRateLimiter(
+	logger log.Logger,
+	metricsClient metrics.Client,
+	domainCache cache.DomainCache,
+	config *config.Config,
+	controller shard.Controller,
+) RateLimiter {
+	rps := func(domain string) int {
+		totalShards := float64(config.NumberOfShards)
+		totalRPS := float64(config.TaskSchedulerGlobalDomainRPS(domain))
+		numShards := float64(controller.NumShards())
+		return int(totalRPS * numShards / totalShards)
+	}
+	limiterFactory := quotas.NewSimpleDynamicRateLimiterFactory(rps)
+	return &taskRateLimiterImpl{
+		logger:           logger,
+		metricsScope:     metricsClient.Scope(metrics.TaskSchedulerRateLimiterScope),
+		domainCache:      domainCache,
+		enabled:          config.TaskSchedulerEnableRateLimiter,
+		enableShadowMode: config.TaskSchedulerEnableRateLimiterShadowMode,
+		limiters:         quotas.NewCollection(limiterFactory),
+	}
+}
+
+func (r *taskRateLimiterImpl) Allow(t Task) bool {
+	if !r.enabled() {
+		return true
+	}
+	limiter, scope, shadow := r.getLimiter(t)
+	allow := limiter.Allow()
+	if allow {
+		return true
+	}
+	scope.IncCounter(metrics.TaskSchedulerThrottledCounterPerDomain)
+	return shadow
+}
+
+func (r *taskRateLimiterImpl) Wait(ctx context.Context, t Task) error {
+	if !r.enabled() {
+		return nil
+	}
+	limiter, scope, shadow := r.getLimiter(t)
+	rsv := limiter.Reserve()
+	allow := rsv.Allow()
+	if allow {
+		rsv.Used(true)
+		return nil
+	}
+	rsv.Used(false)
+	scope.IncCounter(metrics.TaskSchedulerThrottledCounterPerDomain)
+	if shadow {
+		return nil
+	}
+	return limiter.Wait(ctx)
+}
+
+func (r *taskRateLimiterImpl) getLimiter(t Task) (quotas.Limiter, metrics.Scope, bool) {
+	domainID := t.GetDomainID()
+	domainName, err := r.domainCache.GetDomainName(domainID)
+	if err != nil {
+		r.logger.Warn("failed to get domain name from domain cache", tag.WorkflowDomainID(domainID), tag.Error(err))
```

Comment:
What will `r.limiters.For(domainName)` return for non-existing domains? Should we return here while handling err instead of letting the below line execute?

Reply:
for the normal collections: all that matters is if there's an override for that name.  if there is, you get that override.  if not, you get the default value.